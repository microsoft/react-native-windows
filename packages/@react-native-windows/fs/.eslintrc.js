module.exports = {
  extends: ['@rnw-scripts'],
  parserOptions: {tsconfigRootDir : __dirname},
  overrides: [
    {
      files: ['*.ts'],
      // Also not clean
      rules: {
        '@typescript-eslint/no-unused-vars': 'off',
      },
    },
  ],
};
