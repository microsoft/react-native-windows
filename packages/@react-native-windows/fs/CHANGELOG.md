# Change Log - @react-native-windows/fs

This log was last generated on Wed, 27 Apr 2022 05:10:24 GMT and should not be manually modified.

<!-- Start content -->

## 0.0.0-canary.3

Wed, 27 Apr 2022 05:10:24 GMT

### Changes

- Merge (34109996+chiaramooney@users.noreply.github.com)
- Bump @rnw-scripts/eslint-config to v1.1.12

## 0.0.0-canary.2

Tue, 12 Apr 2022 05:10:38 GMT

### Changes

- Fix Overrides (34109996+chiaramooney@users.noreply.github.com)

## 0.0.0-canary.1

Fri, 01 Apr 2022 05:09:57 GMT

### Changes

- Fix publish after promotion to preview (ngerlem@microsoft.com)

## 1.0.2

Wed, 09 Feb 2022 06:09:36 GMT

### Patches

- Bump minimum Node version to 14 (jthysell@microsoft.com)
- Bump @rnw-scripts/eslint-config to v1.1.11
- Bump @rnw-scripts/just-task to v2.2.3
- Bump @rnw-scripts/ts-config to v2.0.2

## 1.0.1

Fri, 03 Dec 2021 06:12:10 GMT

### Patches

- Integrate 11/11 (34109996+chiaramooney@users.noreply.github.com)
- Bump @rnw-scripts/eslint-config to v1.1.10
- Bump @rnw-scripts/just-task to v2.2.2

## 1.0.0

Thu, 18 Nov 2021 06:11:20 GMT

### Major changes

- Add @react-native-windows/fs (ngerlem@microsoft.com)
- Bump @rnw-scripts/eslint-config to v1.1.9
