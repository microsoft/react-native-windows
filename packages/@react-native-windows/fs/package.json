{
  "name": "@react-native-windows/fs",
  "description": "A minimal-dependency drop-in replacement to `fs` with changes for resiliency, promises, and convenience.",
  "version": "0.68.0",
  "license": "MIT",
  "scripts": {
    "build": "rnw-scripts build",
    "clean": "rnw-scripts clean",
    "lint": "rnw-scripts lint",
    "lint:fix": "rnw-scripts lint:fix",
    "watch": "rnw-scripts watch"
  },
  "main": "lib-commonjs/fs.js",
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/react-native-windows.git",
    "directory": "packages/@react-native-windows/fs"
  },
  "dependencies": {
    "graceful-fs": "^4.2.8"
  },
  "devDependencies": {
    "@rnw-scripts/eslint-config": "1.1.10",
    "@rnw-scripts/just-task": "2.2.2",
    "@rnw-scripts/ts-config": "2.0.1",
    "@types/graceful-fs": "^4.1.5",
    "@types/node": "12.12.6",
    "eslint": "^7.32.0",
    "just-scripts": "^1.3.3",
    "prettier": "^2.4.1",
    "typescript": "^4.4.4"
  },
  "files": [
    "assets",
    "lib-commonjs"
  ],
  "engines": {
    "node": ">= 14"
  },
  "promoteRelease": true,
  "beachball": {
    "defaultNpmTag": "latest",
    "disallowedChangeTypes": [
      "major",
      "minor",
      "prerelease"
    ]
  }
}
