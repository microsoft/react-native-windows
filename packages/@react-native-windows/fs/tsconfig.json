{
  "extends": "@rnw-scripts/ts-config",
  "include": ["src"],
  "exclude": ["node_modules"]
}
