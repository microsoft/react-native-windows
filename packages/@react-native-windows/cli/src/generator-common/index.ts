/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 * @format
 */

import fs from 'fs';
import chalk from 'chalk';
import prompts from 'prompts';
import path from 'path';
import mustache from 'mustache';
import {CodedError} from '@react-native-windows/telemetry';

/**
 * Text to replace, + config options
 */
export type Replacements = {
  useMustache?: boolean;
  regExpPatternsToRemove?: RegExp[];
  [key: string]: any;
};

function walk(current: string): string[] {
  if (!fs.lstatSync(current).isDirectory()) {
    return [current];
  }

  const files = fs
    .readdirSync(current)
    .map(child => walk(path.join(current, child)));
  const result: string[] = [];
  return result.concat.apply([current], files);
}

/**
 * Get a source file and replace parts of its contents.
 * @param srcPath Path to the source file.
 * @param replacements e.g. {'TextToBeReplaced': 'Replacement'}
 * @return The contents of the file with the replacements applied.
 */
export function resolveContents(
  srcPath: string,
  replacements: Replacements,
): string {
  let content = fs.readFileSync(srcPath, 'utf8');

  if (content.includes('\r\n')) {
    // CRLF file, make sure multiline replacements are also CRLF
    for (const key of Object.keys(replacements)) {
      if (typeof replacements[key] === 'string') {
        replacements[key] = replacements[key].replace(/(?<!\r)\n/g, '\r\n');
      }
    }
  } else {
    // LF file, make sure multiline replacements are also LF
    for (const key of Object.keys(replacements)) {
      if (typeof replacements[key] === 'string') {
        replacements[key] = replacements[key].replace(/\r\n/g, '\n');
      }
    }
  }

  if (replacements.useMustache) {
    content = mustache.render(content, replacements);
    (replacements.regExpPatternsToRemove || []).forEach(regexPattern => {
      content = content.replace(new RegExp(regexPattern, 'g'), '');
    });
  } else {
    Object.keys(replacements).forEach(regex => {
      content = content.replace(new RegExp(regex, 'g'), replacements[regex]);
    });
  }
  return content;
}

// Binary files, don't process these (avoid decoding as utf8)
const binaryExtensions = ['.png', '.jar', '.keystore'];

type ContentChangedCallbackOption = 'identical' | 'changed' | 'new' | null;

type ContentChangedCallback = (
  path: string,
  option: ContentChangedCallbackOption,
) => Promise<'keep' | 'overwrite'>;

/**
 * Copy a file to given destination, replacing parts of its contents.
 * @param srcPath Path to a file to be copied.
 * @param destPath Destination path.
 * @param replacements: e.g. {'TextToBeReplaced': 'Replacement'}
 * @param contentChangedCallback
 *        Used when upgrading projects. Based on if file contents would change
 *        when being replaced, allows the caller to specify whether the file
 *        should be replaced or not.
 *        If null, files will be overwritten.
 *        Function(path, 'identical' | 'changed' | 'new') => 'keep' | 'overwrite'
 */
export async function copyAndReplace(
  srcPath: string,
  destPath: string,
  replacements: Replacements,
  contentChangedCallback: ContentChangedCallback | null,
) {
  if (fs.lstatSync(srcPath).isDirectory()) {
    if (!fs.existsSync(destPath)) {
      fs.mkdirSync(destPath);
    }
    // Not recursive
    return;
  }

  const extension = path.extname(srcPath);
  if (binaryExtensions.includes(extension)) {
    // Binary file
    let shouldOverwrite = 'overwrite';
    if (contentChangedCallback) {
      const newContentBuffer = fs.readFileSync(srcPath);
      let contentChanged: ContentChangedCallbackOption = 'identical';
      try {
        const origContentBuffer = fs.readFileSync(destPath);
        if (Buffer.compare(origContentBuffer, newContentBuffer) !== 0) {
          contentChanged = 'changed';
        }
      } catch (err) {
        if (err.code === 'ENOENT') {
          contentChanged = 'new';
        } else {
          throw err;
        }
      }
      shouldOverwrite = await contentChangedCallback(destPath, contentChanged);
    }
    if (shouldOverwrite === 'overwrite') {
      copyBinaryFile(srcPath, destPath, err => {
        if (err) {
          throw err;
        }
      });
    }
  } else {
    // Text file
    const srcPermissions = fs.statSync(srcPath).mode;
    const content = resolveContents(srcPath, replacements);

    let shouldOverwrite = 'overwrite';
    if (contentChangedCallback) {
      // Check if contents changed and ask to overwrite
      let contentChanged: ContentChangedCallbackOption = 'identical';
      try {
        const origContent = fs.readFileSync(destPath, 'utf8');
        if (content !== origContent) {
          // logger.info('Content changed: ' + destPath);
          contentChanged = 'changed';
        }
      } catch (err) {
        if (err.code === 'ENOENT') {
          contentChanged = 'new';
        } else {
          throw err;
        }
      }
      shouldOverwrite = await contentChangedCallback(destPath, contentChanged);
    }
    if (shouldOverwrite === 'overwrite') {
      fs.writeFileSync(destPath, content, {
        encoding: 'utf8',
        mode: srcPermissions,
      });
    }
  }
}

/**
 * Same as 'cp' on Unix. Don't do any replacements.
 */
function copyBinaryFile(
  srcPath: string,
  destPath: string,
  cb: (err?: Error) => void,
) {
  let cbCalled = false;
  const srcPermissions = fs.statSync(srcPath).mode;
  const readStream = fs.createReadStream(srcPath);
  readStream.on('error', err => {
    done(err);
  });
  const writeStream = fs.createWriteStream(destPath, {
    mode: srcPermissions,
  });
  writeStream.on('error', err => {
    done(err);
  });
  writeStream.on('close', () => {
    done();
  });
  readStream.pipe(writeStream);
  function done(err?: Error) {
    if (!cbCalled) {
      cb(err);
      cbCalled = true;
    }
  }
}

export function createDir(destPath: string) {
  if (!fs.existsSync(destPath)) {
    fs.mkdirSync(destPath);
  }
}

export async function copyAndReplaceWithChangedCallback(
  srcPath: string,
  destRoot: string,
  relativeDestPath: string,
  replacements?: Record<string, string>,
  alwaysOverwrite?: boolean,
) {
  if (!replacements) {
    replacements = {};
  }
  const contentChangedCallback: ContentChangedCallback = alwaysOverwrite
    ? (_, contentChanged) =>
        alwaysOverwriteContentChangedCallback(
          srcPath,
          relativeDestPath,
          contentChanged,
        )
    : (_, contentChanged) =>
        upgradeFileContentChangedCallback(
          srcPath,
          relativeDestPath,
          contentChanged,
        );

  await copyAndReplace(
    srcPath,
    path.join(destRoot, relativeDestPath),
    replacements,
    contentChangedCallback,
  );
}

export async function copyAndReplaceAll(
  srcPath: string,
  destPath: string,
  relativeDestDir: string,
  replacements: Replacements,
  alwaysOverwrite: boolean,
) {
  for (const absoluteSrcFilePath of walk(srcPath)) {
    const filename = path.relative(srcPath, absoluteSrcFilePath);
    const relativeDestPath = path.join(relativeDestDir, filename);
    await copyAndReplaceWithChangedCallback(
      absoluteSrcFilePath,
      destPath,
      relativeDestPath,
      replacements,
      alwaysOverwrite,
    );
  }
}

async function alwaysOverwriteContentChangedCallback(
  absoluteSrcFilePath: string,
  relativeDestPath: string,
  contentChanged: ContentChangedCallbackOption,
): Promise<'keep' | 'overwrite'> {
  if (contentChanged === 'new') {
    console.log(`${chalk.bold('new')} ${relativeDestPath}`);
    return 'overwrite';
  }
  if (contentChanged === 'changed') {
    console.log(
      `${chalk.bold('changed')} ${relativeDestPath} ${chalk.yellow(
        '[overwriting]',
      )}`,
    );
    return 'overwrite';
  }
  if (contentChanged === 'identical') {
    return 'keep';
  }
  throw new CodedError(
    'Autolinking',
    `Unknown file changed state: ${relativeDestPath}, ${contentChanged}`,
  );
}

async function upgradeFileContentChangedCallback(
  absoluteSrcFilePath: string,
  relativeDestPath: string,
  contentChanged: ContentChangedCallbackOption,
): Promise<'keep' | 'overwrite'> {
  if (contentChanged === 'new') {
    console.log(`${chalk.bold('new')} ${relativeDestPath}`);
    return 'overwrite';
  }
  if (contentChanged === 'changed') {
    console.log(
      `${chalk.bold(relativeDestPath)} ` +
        `has changed in the new version.\nDo you want to keep your ${relativeDestPath} or replace it with the ` +
        'latest version?\nMake sure you have any changes you made to this file saved somewhere.\n' +
        `You can see the new version here: ${absoluteSrcFilePath}`,
    );

    const {shouldReplace} = await prompts([
      {
        name: 'shouldReplace',
        type: 'confirm',
        message: `Do you want to replace ${relativeDestPath}?`,
        initial: false,
      },
    ]);

    return shouldReplace ? 'overwrite' : 'keep';
  }
  if (contentChanged === 'identical') {
    return 'keep';
  }
  throw new CodedError(
    'Autolinking',
    `Unknown file changed state: ${relativeDestPath}, ${contentChanged}`,
  );
}
