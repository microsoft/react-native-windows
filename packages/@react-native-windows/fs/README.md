# @react-native-windows/fs

`@react-native-windows/fs` is a minimal-dependency drop-in replacement to `fs` with changes for
resiliency, promises, and convenience. It has several opinionated changes, targeted towards CLI
applications handling JavaScript-oriented files.

![Usage Thumbnail](./assets/logo.png)

## Async Usage (Default)

`@react-native-windows/fs` exposes a Promise-based API, mostly matching that of `fs.promises`, with
several methods added [extra methods](#Extra-Methods).

```ts
// import {promises as fs} from 'fs'
import fs from '@react-native-windows/fs';

const fileContent = await fs.readFile('foo.txt');
```

## Sync Usage

`@react-native-windows/fs` exports all `fs.*Sync` Where an async version has a graceful
implementation, and the synchronous version does not, the method is marked as deprecated.

```ts
// import fs from 'fs'
import fs from '@react-native-windows/fs';

const fileContent = fs.readFileSync('foo.txt');
```

## Extra Methods

### `exists`
NodeJS deprecated `fs.exists`, and removed `fs.promises.exists`. The recommendation is to instead
acquire a lock to the file via `fs.open` for the duration of file-use. One-shot existence checks are
still useful, and because `fs.existsSync`is not deprecated, more likely means usage of blocking
synchronous APIs.

```ts
import fs from '@react-native-windows/fs';

const fooExists = await fs.exists('foo.txt');
```

### `readJsonFile` and `readJsonFileSync`

`@react-native-windows/fs` provides convenience methods to handle JSON files. The following methods
are added:

| Method | Return type |
|-|-|
| `readJsonFile<T>` | `Promise<T>` or `Promise<Record<string, unknown>>` |
| `readJsonFileSync<T>` | `T` or `Record<string, unknown>` |

```ts
import fs from '@react-native-windows/fs';

// foo is type: Record<string, unknown> by default
const foo = await fs.readJsonFile('foo.json');

// foo is type: FooProps
type FooProps = { name: string, version: string };
const foo = await fs.readJsonFile<FooProps>('foo.json');
```

## Resiliency

`@react-native-windows/fs` uses [`graceful-fs`](https://github.com/isaacs/node-graceful-fs) to
gracefully handle transient filesystem conditions, at the cost of extra latency. This includes
transient `EPERM`, `EACCESS`, `EMFILE`, `ENFILE`. This can be important when handling files that a
subject to antivirus, which may temporarily lock mutation of files on Windows.

## eslint
We reccomend adding the following rules to your eslint config if you would like to use
`@react-native-windows/fs` everywhere:
```js
module.exports = {
  rules: {
    'no-restricted-imports': [
      'error', {
        name: 'fs',
        message: 'Please use `@react-native-windows/fs` instead of `fs`'
      }
    ],
  }
}
```
