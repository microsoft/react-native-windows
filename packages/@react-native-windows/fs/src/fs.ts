/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * @format
 */

import asyncMethods, {AsyncMethods} from './asyncMethods';
import syncMethods, {SyncMethods} from './syncMethods';
import fs from 'graceful-fs';

export interface ReactBativeWindowsFs extends AsyncMethods, SyncMethods {
  constants: typeof fs.constants;

  /**
   * `renameSync` should not be used, as it can cause EPERM issues on Windows
   * @deprecated
   */
  renameSync(oldPath: fs.PathLike, newPath: fs.PathLike): void;
}

const rnwFs: ReactBativeWindowsFs = {
  constants: fs.constants,
  ...asyncMethods,
  ...syncMethods,
};
export default rnwFs;
