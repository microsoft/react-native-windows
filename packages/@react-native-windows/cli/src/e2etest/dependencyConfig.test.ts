/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 * @format
 */

import fs from 'fs';
import path from 'path';

import {
  dependencyConfigWindows,
  WindowsDependencyConfig,
} from '../config/dependencyConfig';

import {copyAndReplace} from '../generator-common';

const templateRoot = path.resolve('../../../vnext/template');

const testProjectGuid = '{416476D5-974A-4EE2-8145-4E331297247E}';

async function tryMkdir(dir: string): Promise<void> {
  try {
    await fs.promises.mkdir(dir, {recursive: true});
  } catch (err) {}
}

type TargetProject = [string, ((folder: string) => Promise<void>)?];

function project(
  name: string,
  setup?: (folder: string) => Promise<void>,
): TargetProject {
  return [name, setup];
}

const projects: TargetProject[] = [
  // Nothing but a react-native.config.js, with nulls
  project('BlankLib'),
  // Nothing but a windows folder
  project('MissingProjectFilesLib', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);
  }),
  // New C++ app project based on the template
  project('SimpleCppApp', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);

    const replacements = {
      name: 'SimpleCppApp',
      namespace: 'SimpleCppApp',
      useMustache: true,
      projectGuidUpper: testProjectGuid,
      projectGuidLower: testProjectGuid.toLowerCase(),
    };

    await copyAndReplace(
      path.join(templateRoot, 'cpp-app/proj/MyApp.sln'),
      path.join(windowsDir, 'SimpleCppApp.sln'),
      replacements,
      null,
    );

    const projDir = path.join(windowsDir, 'SimpleCppApp');
    await tryMkdir(projDir);

    await copyAndReplace(
      path.join(templateRoot, 'cpp-app/proj/MyApp.vcxproj'),
      path.join(projDir, 'SimpleCppApp.vcxproj'),
      replacements,
      null,
    );
  }),
  // New C++ project based on the template
  project('SimpleCppLib', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);

    const replacements = {
      name: 'SimpleCppLib',
      namespace: 'SimpleCppLib',
      useMustache: true,
      projectGuidUpper: testProjectGuid,
      projectGuidLower: testProjectGuid.toLowerCase(),
    };

    await copyAndReplace(
      path.join(templateRoot, 'cpp-lib/proj/MyLib.sln'),
      path.join(windowsDir, 'SimpleCppLib.sln'),
      replacements,
      null,
    );

    const projDir = path.join(windowsDir, 'SimpleCppLib');
    await tryMkdir(projDir);

    await copyAndReplace(
      path.join(templateRoot, 'cpp-lib/proj/MyLib.vcxproj'),
      path.join(projDir, 'SimpleCppLib.vcxproj'),
      replacements,
      null,
    );
  }),
  // New C# app project based on the template
  project('SimpleCSharpApp', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);

    const replacements = {
      name: 'SimpleCSharpApp',
      namespace: 'SimpleCSharpApp',
      useMustache: true,
      projectGuidUpper: testProjectGuid,
      projectGuidLower: testProjectGuid.toLowerCase(),
    };

    await copyAndReplace(
      path.join(templateRoot, 'cs-app/proj/MyApp.sln'),
      path.join(windowsDir, 'SimpleCSharpApp.sln'),
      replacements,
      null,
    );

    const projDir = path.join(windowsDir, 'SimpleCSharpApp');
    await tryMkdir(projDir);

    await copyAndReplace(
      path.join(templateRoot, 'cs-app/proj/MyApp.csproj'),
      path.join(projDir, 'SimpleCSharpApp.csproj'),
      replacements,
      null,
    );
  }),
  // New C# project based on the template
  project('SimpleCSharpLib', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);

    const replacements = {
      name: 'SimpleCSharpLib',
      namespace: 'SimpleCSharpLib',
      useMustache: true,
      projectGuidUpper: testProjectGuid,
      projectGuidLower: testProjectGuid.toLowerCase(),
    };

    await copyAndReplace(
      path.join(templateRoot, 'cs-lib/proj/MyLib.sln'),
      path.join(windowsDir, 'SimpleCSharpLib.sln'),
      replacements,
      null,
    );

    const projDir = path.join(windowsDir, 'SimpleCSharpLib');
    await tryMkdir(projDir);

    await copyAndReplace(
      path.join(templateRoot, 'cs-lib/proj/MyLib.csproj'),
      path.join(projDir, 'SimpleCSharpLib.csproj'),
      replacements,
      null,
    );
  }),
  project('WithIndirectDependency'),
];

// Tests that given userConfig is null, the result will always be null
test.each(projects)(
  'dependencyConfig - %s (userConfig is null)',
  async (name, setup) => {
    const folder = path.resolve('src/e2etest/projects/', name);

    if (setup !== undefined) {
      await setup(folder);
    }

    const userConfig = null;
    const expectedConfig: WindowsDependencyConfig | null = null;

    expect(dependencyConfigWindows(folder, userConfig)).toBe(expectedConfig);
  },
);

// Tests the result given a windows project config in react-native.config.js
test.each(projects)(
  'dependencyConfig - %s (Use react-native.config.js)',
  async (name, setup) => {
    const folder = path.resolve('src/e2etest/projects/', name);
    const rnc = require(path.join(folder, 'react-native.config.js'));

    if (setup !== undefined) {
      await setup(folder);
    }

    const userConfig: Partial<WindowsDependencyConfig> =
      rnc.dependency.platforms.windows;

    if (name === 'BlankLib') {
      expect(dependencyConfigWindows(folder, userConfig)).toMatchSnapshot();
    } else if (name.endsWith('App')) {
      expect(dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
        folder: expect.stringContaining(name),
        projects: expect.arrayContaining([
          expect.objectContaining({
            projectFile: expect.stringMatching(
              /Error: .*\.(?:cs|vcx)proj is type '\w+'/,
            ),
          }),
        ]),
      });
    } else {
      expect(dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
        folder: expect.stringContaining(name),
      });
    }
  },
);

// Tests the result of ignoring the windows project config in react-native.config.js
test.each(projects)(
  'dependencyConfig - %s (Ignore react-native.config.js)',
  async (name, setup) => {
    const folder = path.resolve('src/e2etest/projects/', name);

    if (setup !== undefined) {
      await setup(folder);
    }

    const userConfig: Partial<WindowsDependencyConfig> = {};

    if (name === 'BlankLib') {
      expect(dependencyConfigWindows(folder, userConfig)).toMatchSnapshot();
    } else if (name.endsWith('App')) {
      expect(dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
        folder: expect.stringContaining(name),
        projects: expect.arrayContaining([
          expect.objectContaining({
            projectFile: expect.stringMatching(
              /Error: .*\.(?:cs|vcx)proj is type '\w+'/,
            ),
          }),
        ]),
      });
    } else {
      expect(dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
        folder: expect.stringContaining(name),
      });
    }
  },
);
