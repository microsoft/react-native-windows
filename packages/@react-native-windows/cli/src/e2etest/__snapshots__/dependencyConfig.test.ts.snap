// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependencyConfig - BlankLib (Ignore react-native.config.js) 1`] = `null`;

exports[`dependencyConfig - BlankLib (Use react-native.config.js) 1`] = `null`;

exports[`dependencyConfig - MissingProjectFilesLib (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "MissingProjectFilesLib",
  "nugetPackages": Array [],
  "projects": Array [],
  "solutionFile": null,
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - MissingProjectFilesLib (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "MissingProjectFilesLib",
  "nugetPackages": Array [],
  "projects": Array [],
  "solutionFile": null,
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCSharpApp (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCSharpApp",
  "nugetPackages": Array [],
  "projects": ArrayContaining [
    ObjectContaining {
      "projectFile": StringMatching /Error: \\.\\*\\\\\\.\\(\\?:cs\\|vcx\\)proj is type '\\\\w\\+'/,
    },
  ],
  "solutionFile": "SimpleCSharpApp.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCSharpApp (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCSharpApp",
  "nugetPackages": Array [],
  "projects": ArrayContaining [
    ObjectContaining {
      "projectFile": StringMatching /Error: \\.\\*\\\\\\.\\(\\?:cs\\|vcx\\)proj is type '\\\\w\\+'/,
    },
  ],
  "solutionFile": "SimpleCSharpApp.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCSharpLib (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCSharpLib",
  "nugetPackages": Array [],
  "projects": Array [
    Object {
      "cppHeaders": Array [
        "winrt/SimpleCSharpLib.h",
      ],
      "cppPackageProviders": Array [
        "SimpleCSharpLib::ReactPackageProvider",
      ],
      "csNamespaces": Array [
        "SimpleCSharpLib",
      ],
      "csPackageProviders": Array [
        "SimpleCSharpLib.ReactPackageProvider",
      ],
      "directDependency": true,
      "projectFile": "SimpleCSharpLib\\\\SimpleCSharpLib.csproj",
      "projectGuid": "{416476D5-974A-4EE2-8145-4E331297247E}",
      "projectLang": "cs",
      "projectName": "SimpleCSharpLib",
    },
  ],
  "solutionFile": "SimpleCSharpLib.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCSharpLib (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCSharpLib",
  "nugetPackages": Array [],
  "projects": Array [
    Object {
      "cppHeaders": Array [
        "winrt/SimpleCSharpLib.h",
      ],
      "cppPackageProviders": Array [
        "SimpleCSharpLib::ReactPackageProvider",
      ],
      "csNamespaces": Array [
        "SimpleCSharpLib",
      ],
      "csPackageProviders": Array [
        "SimpleCSharpLib.ReactPackageProvider",
      ],
      "directDependency": true,
      "projectFile": "SimpleCSharpLib\\\\SimpleCSharpLib.csproj",
      "projectGuid": "{416476D5-974A-4EE2-8145-4E331297247E}",
      "projectLang": "cs",
      "projectName": "SimpleCSharpLib",
    },
  ],
  "solutionFile": "SimpleCSharpLib.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCppApp (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCppApp",
  "nugetPackages": Array [],
  "projects": ArrayContaining [
    ObjectContaining {
      "projectFile": StringMatching /Error: \\.\\*\\\\\\.\\(\\?:cs\\|vcx\\)proj is type '\\\\w\\+'/,
    },
  ],
  "solutionFile": "SimpleCppApp.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCppApp (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCppApp",
  "nugetPackages": Array [],
  "projects": ArrayContaining [
    ObjectContaining {
      "projectFile": StringMatching /Error: \\.\\*\\\\\\.\\(\\?:cs\\|vcx\\)proj is type '\\\\w\\+'/,
    },
  ],
  "solutionFile": "SimpleCppApp.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCppLib (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCppLib",
  "nugetPackages": Array [],
  "projects": Array [
    Object {
      "cppHeaders": Array [
        "winrt/SimpleCppLib.h",
      ],
      "cppPackageProviders": Array [
        "SimpleCppLib::ReactPackageProvider",
      ],
      "csNamespaces": Array [
        "SimpleCppLib",
      ],
      "csPackageProviders": Array [
        "SimpleCppLib.ReactPackageProvider",
      ],
      "directDependency": true,
      "projectFile": "SimpleCppLib\\\\SimpleCppLib.vcxproj",
      "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
      "projectLang": "cpp",
      "projectName": "SimpleCppLib",
    },
  ],
  "solutionFile": "SimpleCppLib.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - SimpleCppLib (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCppLib",
  "nugetPackages": Array [],
  "projects": Array [
    Object {
      "cppHeaders": Array [
        "winrt/SimpleCppLib.h",
      ],
      "cppPackageProviders": Array [
        "SimpleCppLib::ReactPackageProvider",
      ],
      "csNamespaces": Array [
        "SimpleCppLib",
      ],
      "csPackageProviders": Array [
        "SimpleCppLib.ReactPackageProvider",
      ],
      "directDependency": true,
      "projectFile": "SimpleCppLib\\\\SimpleCppLib.vcxproj",
      "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
      "projectLang": "cpp",
      "projectName": "SimpleCppLib",
    },
  ],
  "solutionFile": "SimpleCppLib.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - WithIndirectDependency (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "WithIndirectDependency",
  "nugetPackages": Array [],
  "projects": Array [],
  "solutionFile": "WithIndirectDependency.sln",
  "sourceDir": "windows",
}
`;

exports[`dependencyConfig - WithIndirectDependency (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "WithIndirectDependency",
  "nugetPackages": Array [],
  "projects": Array [
    Object {
      "cppHeaders": Array [
        "winrt/WithIndirectDependency.h",
      ],
      "cppPackageProviders": Array [
        "WithIndirectDependency::ReactPackageProvider",
      ],
      "csNamespaces": Array [
        "WithIndirectDependency",
      ],
      "csPackageProviders": Array [
        "WithIndirectDependency.ReactPackageProvider",
      ],
      "directDependency": true,
      "projectFile": "WithIndirectDependency\\\\WithIndirectDependency.vcxproj",
      "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
      "projectLang": "cpp",
      "projectName": "WithIndirectDependency",
    },
    Object {
      "directDependency": false,
      "projectFile": "IndirectDependency\\\\IndirectDependency.vcxproj",
      "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
      "projectLang": "cpp",
      "projectName": "IndirectDependency",
    },
  ],
  "solutionFile": "WithIndirectDependency.sln",
  "sourceDir": "windows",
}
`;
